src/main/java/io/galeb/manager/entity/service/EtagService.java

```diff
@@ -72,151 +58,94 @@ public EtagService(final EnvironmentRepository environmentRepository,
+    private void updateLastEtag(String envname, String etagChanges) {
+        template.opsForHash().put(getInfoKey(envname), FIELD_INFO_ETAG, etagChanges);
     }
 
-    private String getBodyCached(String etag, String envname) {
-        if (template.hasKey(getEtagKey(etag, envname))) {
-            return Optional.ofNullable(template.opsForValue().get(getEtagKey(etag, envname))).orElse("");
-        }
-        return "";
+    private boolean isAllEtagEquals(String etagChanges, String lastETag, String routerEtag) {
+        return etagChanges.equals(lastETag) && etagChanges.equals(routerEtag);
     }
 
-    private String getEtagKey(String etag, String envname) {
-        return PREFIX_ETAG + ":" + envname + ":" + etag;
+    private Set<String> changes(String envname) {
+        final Set<String> result = template.keys(PREFIX_HAS_CHANGE + envname + ":*");
+        return (result != null) ? result : Collections.emptySet();
     }
 
-    private String newEtag(String envname, int numRouters) {
-        try {
-            List<VirtualHost> virtualHosts = copyService.getVirtualHosts(envname, numRouters);
-            if (virtualHosts == null || virtualHosts.isEmpty()) {
-                return "";
-            }
-            String key = virtualHosts.stream().map(this::getFullHash)
-                    .sorted()
-                    .distinct()
-                    .collect(Collectors.joining());
-            String etag = key == null || "".equals(key) ? "" : sha256().hashString(key, Charsets.UTF_8).toString();
-            virtualHosts = virtualHosts.stream()
-                    .map(v -> {
-                        v.getEnvironment().getProperties().put(PROP_FULLHASH, etag);
-                        return v;
-                    })
-                    .collect(Collectors.toList());
-            persistToDb(envname, etag);
-            persistToRedis(etag, envname, gson.toJson(new Virtualhosts(virtualHosts.toArray(new VirtualHost[]{})), Virtualhosts.class));
-            return etag;
-        } catch (Exception e) {
-            ErrorLogger.logError(e, this.getClass());
-        }
-        return "";
+    private void persistToRedis(String etag, String envname, String body) {
+        Map<String, String> mapValues = new HashMap<String, String>() {{
+            put(FIELD_INFO_CACHE, body);
+            put(FIELD_INFO_ETAG, etag);
+        }};
+        template.opsForHash().putAll(getInfoKey(envname), mapValues);
     }
 
-    @Transactional
-    private void persistToDb(String envname, String etag) throws Exception {
-        Authentication currentUser = CurrentUser.getCurrentAuth();
-        SystemUserService.runAs();
-        updateEtag(envFindByName(envname), etag);
-        SystemUserService.runAs(currentUser);
+    private String getLastEtag(String envname) {
+        String valueETag = (String)template.opsForHash().get(getInfoKey(envname), FIELD_INFO_ETAG);
+        return valueETag != null ? valueETag : RandomStringUtils.random(10);
```

Comment:
Why using org.apache.commons.lang.RandomStringUtils? java.util.UUID is builtin JDK, avoiding external dependence